Give URLKeepingBlobAlive a top-origin SecurityOriginData

sysrqb opened the pull request:
#### 6f0e501e8e36d638026626ed2051d73db57c8f88
<pre>
Give URLKeepingBlobAlive a top-origin SecurityOriginData
<a href="https://bugs.webkit.org/show_bug.cgi?id=251218">https://bugs.webkit.org/show_bug.cgi?id=251218</a>
rdar://104704879

Reviewed by Chris Dumez.

In this patch we construct the URLKeepingBlobAlive with its associated
top-level SecurityOrigin(Data). This will be helpful in a future patch where we
partition the Blob registry using that top-level origin. URLKeepingBlobAlive
needs this information, in particular, because it is responsible for
automatically registering and unregistering its Blob URL handle - thus keeping
the blob alive as long as it exists, as the class name implies.

* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::FetchRequest::FetchRequest):
(WebCore::m_signal):
(WebCore::FetchRequest::initializeWith):
(WebCore::FetchRequest::create):
(WebCore::FetchRequest::clone):
(WebCore::FetchRequest::stop):
* Source/WebCore/Modules/fetch/FetchRequest.h:
(WebCore::FetchRequest::FetchRequest): Deleted.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setURL):
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::Blob::Blob):
(WebCore::Blob::handle const):
* Source/WebCore/fileapi/Blob.h:
* Source/WebCore/fileapi/URLKeepingBlobAlive.cpp:
(WebCore::URLKeepingBlobAlive::URLKeepingBlobAlive):
(WebCore::URLKeepingBlobAlive::clear):
(WebCore::URLKeepingBlobAlive::operator=):
(WebCore::URLKeepingBlobAlive::isolatedCopy const):
(WebCore::URLKeepingBlobAlive::isolatedCopy):
* Source/WebCore/fileapi/URLKeepingBlobAlive.h:
(WebCore::URLKeepingBlobAlive::URLKeepingBlobAlive): Deleted.
(WebCore::URLKeepingBlobAlive::operator=): Deleted.
* Source/WebCore/loader/NavigationScheduler.cpp:
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::extendBlobURLLifetimeIfNecessary const):
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):
* Source/WebCore/loader/PolicyChecker.h:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::importScripts):
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::SharedWorker::SharedWorker):
(WebCore::SharedWorker::didFinishLoading):
* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::open):
(WebCore::XMLHttpRequest::createRequest):
(WebCore::XMLHttpRequest::clearRequest):
(WebCore::XMLHttpRequest::didFinishLoading):

Canonical link: <a href="https://commits.webkit.org/260266@main">https://commits.webkit.org/260266@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54ff2167c84946efd9a28455bccb3e691195ccd4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16622 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40466 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116724 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116139 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111460 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18047 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7937 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99732 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113324 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13649 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96817 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41288 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95560 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28443 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83080 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9621 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29796 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10284 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6693 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15778 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49377 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11846 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3849 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -184,6 +191,7 @@ Blob::Blob(ReferencingExistingBlobConstructor, ScriptExecutionContext* context,
     , m_size(blob.size())
     , m_memoryCost(blob.memoryCost())
     , m_internalURL(BlobURL::createInternalURL())
+    , m_topOrigin(context ? context->topOrigin().data() : SecurityOriginData::createOpaque())
```
Do we really need this data member? Couldn't we just call `scriptExecutionContext()->topOrigin().data()` whenever we need a topOrigin? We could just fallback to an opaqueOrigin if `scriptExecutionContext()` returns null, just like you do here.

sysrqb replied:
Ideally, we would not need this, but I found that caching the SecurityOriginData was needed here while I worked on #8258 because the ScriptExecutionContext could be destroyed before the Blob is destroyed (and we'll need it when we unregister). So, even though we don't need this additional data member in this patch, it will be helpful in a follow-up. I can change this like you described and add the data member in the follow up patch, though, if you prefer.

cdumez replied:
I think this is good enough justification. I was just curious if we could do away with it or not.
Fine to include in this patch given the reason.

cdumez commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -118,14 +119,18 @@ Blob::Blob(UninitializedContructor, ScriptExecutionContext* context, URL&& url,
     , m_internalURL(BlobURL::createInternalURL())
+    , m_topOrigin(context ? context->topOrigin().data() : SecurityOriginData::createOpaque())
 {
+    ASSERT(context);
```
I don't think we should have the assertion, not sure what makes this a guarantee

sysrqb replied:
I didn't find any call sites where the ScriptExecutionContext should be null, so this was only a sanity check.

cdumez replied:
If call sites never passed null, we would take in a C++ reference, not a pointer. If something is a pointer in WebKit, always assume it can be null. If it really cannot, then update the parameter to be a c++ reference.

cdumez commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -118,14 +119,18 @@ Blob::Blob(UninitializedContructor, ScriptExecutionContext* context, URL&& url,
     , m_internalURL(WTFMove(url))
+    , m_topOrigin(context ? context->topOrigin().data() : SecurityOriginData::createOpaque())
 {
+    ASSERT(context);
```
I don't think we should have the assertion, not sure what makes this a guarantee

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3542,7 +3542,9 @@ void Document::setURL(const URL& url)
         newURL.setHostAndPort({ });
-    m_url = WTFMove(newURL);
+    // SecurityContext::securityOrigin may not be initialized at this time if setURL() is called in the constructor, therefore calling topOrigin() is not always safe.
+    auto topOrigin = isTopDocument() && !SecurityContext::securityOrigin() ? SecurityOriginData::fromURL(url) : this->topOrigin().data();
```
What if the URL is a blob URL? (which is the base we care about I believe).
I see 2 likely issues:
1. SecurityOriginData::fromURL(url) likely won't be able to extra the origin from a blob URL
2. Even if we addressed the first issue, how would it enforce partitioning?

Imagine this case:
- Frame from origin B (under top level origin A) creates a Blob and gets a URL for it. Then it opens a new window and uses this blob URL as URL for the popup. What's supposed to happen? I imagine the blob should have the "A / B" partitioned origin. Seems like we shouldn't be able to load this Blob in a top level popup? If so, I worry that your code here is not going to achieve this (even if we fix issue 1). Am I missing something?

sysrqb replied:
Right, I originally used SecurityOrigin::create(url) for this but I change it in this patch, and that is wrong - we do need SecurityOrigin's logic for handling blob URLs. But, as for (2), there are two options:
1. the popup has "A / B" blob partitioning (therefore has access to the blob data)
2. the popup has "B" blob partitioning (therefore doesn't have access to the blob data)

I was following (2) because creating a top-level navigation with subframe partitioning was conceptually confusing and required more invasive code changes. Following (1) will provide better web compat, but I don't know how common is that use case. If you think we should use (1), then this patch will need changes.

cdumez replied:
I think 2 makes sense. The popup would have "B" origin and thus shouldn't be able to load the blob.

sysrqb replied:
Okay, thanks. I also just remembered and realized you are missing some other context. In https://github.com/w3c/FileAPI/issues/153 there is discussion about how to partition blobs, and the general agreement is that cross-origin top-level navigations to a Blob URL should enforce no-opener.

cdumez reviewed (commented):
I am not opposed to adding a topOrigin to this class (since this class has Blob in its name). However, I have some other comments.

cdumez commented on Source/WebCore/Modules/fetch/FetchRequest.cpp:
```diff
@@ -128,6 +128,18 @@ static bool methodCanHaveBody(const ResourceRequest& request)
+inline FetchRequest::FetchRequest(ScriptExecutionContext* context, std::optional<FetchBody>&& body, Ref<FetchHeaders>&& headers, ResourceRequest&& request, FetchOptions&& options, String&& referrer)
+    : FetchBodyOwner(context, WTFMove(body), WTFMove(headers))
+    , m_request(WTFMove(request))
+    , m_requestURL({ m_request.url(), context->topOrigin().data() })
```
We shouldn't dereference context without a null check. If context really truly can't be null, then we should change the parameter type to be a ScriptExecutionContext&.

sysrqb replied:
Changing the parameter type seemed safe here, so I went ahead and did that.

cdumez reviewed (approved):
LGTM

sysrqb committed:
Give URLKeepingBlobAlive a top-origin SecurityOriginData
https://bugs.webkit.org/show_bug.cgi?id=251218
rdar://104704879

Reviewed by Chris Dumez.

In this patch we construct the URLKeepingBlobAlive with its associated
top-level SecurityOrigin(Data). This will be helpful in a future patch where we
partition the Blob registry using that top-level origin. URLKeepingBlobAlive
needs this information, in particular, because it is responsible for
automatically registering and unregistering its Blob URL handle - thus keeping
the blob alive as long as it exists, as the class name implies.

* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::FetchRequest::FetchRequest):
(WebCore::m_signal):
(WebCore::FetchRequest::initializeWith):
(WebCore::FetchRequest::create):
(WebCore::FetchRequest::clone):
(WebCore::FetchRequest::stop):
* Source/WebCore/Modules/fetch/FetchRequest.h:
(WebCore::FetchRequest::FetchRequest): Deleted.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setURL):
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::Blob::Blob):
(WebCore::Blob::handle const):
* Source/WebCore/fileapi/Blob.h:
* Source/WebCore/fileapi/URLKeepingBlobAlive.cpp:
(WebCore::URLKeepingBlobAlive::URLKeepingBlobAlive):
(WebCore::URLKeepingBlobAlive::clear):
(WebCore::URLKeepingBlobAlive::operator=):
(WebCore::URLKeepingBlobAlive::isolatedCopy const):
(WebCore::URLKeepingBlobAlive::isolatedCopy):
* Source/WebCore/fileapi/URLKeepingBlobAlive.h:
(WebCore::URLKeepingBlobAlive::URLKeepingBlobAlive): Deleted.
(WebCore::URLKeepingBlobAlive::operator=): Deleted.
* Source/WebCore/loader/NavigationScheduler.cpp:
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::extendBlobURLLifetimeIfNecessary const):
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):
* Source/WebCore/loader/PolicyChecker.h:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::importScripts):
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::SharedWorker::SharedWorker):
(WebCore::SharedWorker::didFinishLoading):
* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::open):
(WebCore::XMLHttpRequest::createRequest):
(WebCore::XMLHttpRequest::clearRequest):
(WebCore::XMLHttpRequest::didFinishLoading):

Canonical link: https://commits.webkit.org/260266@main
